### PR TITLE
Convert ProtoParser to Kotlin

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt [hunk 1]
```diff
@@ -13,403 +13,402 @@
...
-   * between this and {@code readChar} is that this doesn't consume the char.
+   * Peeks a non-whitespace character and returns it. The only difference between this and
+   * [readChar] is that this doesn't consume the char.
    */
-  public char peekChar() {
-    skipWhitespace(true);
-    if (pos == data.length) throw unexpected("unexpected end of file");
-    return data[pos];
+  fun peekChar(): Char {
+    skipWhitespace(true)
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt]: nit: Name the arg

File: wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt [hunk 2]
```diff
@@ -13,403 +13,402 @@
...
-        tag = tag.substring("0x".length());
-        radix = 16;
+        tag = tag.substring("0x".length)
+        radix = 16
       }
-      return Integer.valueOf(tag, radix);
-    } catch (Exception e) {
-      throw unexpected("expected an integer but was " + tag);
+      return Integer.valueOf(tag, radix)
+    } catch (e: Exception) {
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt]: nit: `_: Exception`

File: wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt [hunk 3]
```diff
@@ -13,403 +13,402 @@
...
   }
 
-  public Location location() {
-    return location.at(line + 1, pos - lineStart + 1);
-  }
+  fun location() = location.at(line + 1, pos - lineStart + 1)
 
-  public RuntimeException unexpected(String message) {
-    return unexpected(location(), message);
+  inline fun expect(condition: Boolean, location: Location = location(), message: () -> String) {
```
Comment (Egorand) [inline wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt]: This is very nice!

File: wire-schema/src/test/java/com/squareup/wire/schema/internal/parser/ProtoParserTest.kt [hunk 4]
```diff
@@ -512,7 +512,7 @@ class ProtoParserTest {
       ProtoParser.parse(location, proto)
     } catch (e: IllegalStateException) {
       assertThat(e).hasMessage(
-          "Syntax error in file.proto at 2:12: expected '//' or '/*'"
+          "Syntax error in file.proto at 2:13: expected '//' or '/*'"
```
Comment (Egorand) [inline wire-schema/src/test/java/com/squareup/wire/schema/internal/parser/ProtoParserTest.kt]: Bug?

File: wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt (hunk 1 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt]: Done x 4.

File: wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt (hunk 2 above)
Comment (swankjesse) [reply wire-schema/src/main/java/com/squareup/wire/schema/internal/parser/SyntaxReader.kt]: Fixed.

File: wire-schema/src/test/java/com/squareup/wire/schema/internal/parser/ProtoParserTest.kt (hunk 4 above)
Comment (swankjesse) [reply wire-schema/src/test/java/com/squareup/wire/schema/internal/parser/ProtoParserTest.kt]: When I converted it from an `if` to a lambda I started sampling the location earlier. That impacts this one callsite.
```
    expect(pos < data.size && (data[pos] == '/' || data[pos] == '*')) {
      pos-- // Backtrack to start of comment.
      "expected '//' or '/*'"
    }
```
I didn’t want to make things more complex to support it, so I changed the test. Both offsets are okay!

Comment (Egorand) [reply wire-schema/src/test/java/com/squareup/wire/schema/internal/parser/ProtoParserTest.kt]: Sweet! Thought so but double-checked.

**Pull request merged**